### PR TITLE
Preallocate for `FixedSizeList` in `concat`

### LINKED ISSUE #4897: concat should preallocate for FixedSizeList
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

Right now, concat only pre-allocates for string, binary, dictionary and primitive arrays (IIUC):

https://github.com/apache/arrow-rs/blob/208da03979b2903c3182c20ef382b2895756380a/arrow-select/src/concat.rs#L129-L139

We use `FixedSizeList<{FloatType}>` as a storage type for tensor arrays, and not pre-allocating can get very expensive in plans. https://github.com/lancedb/lance/issues/1360

**Describe the solution you'd like**

Should provide capacities for fixed size list array. We likely need to add an entry to this enum:

https://github.com/apache/arrow-rs/blob/b46ea46aa65149fac763671b0adcb9c4e406ec11/arrow-data/src/transform/mod.rs#L323-L344

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by judahrand**
# Which issue does this PR close?

Closes #4897

# Rationale for this change
 
This is my first contribution to the project and also my first contribution in Rust so I'd appreciate all input and/or pointers! 

These changes aim to make it so that memory is pre-allocated the values array of `FixedListArray`s which are concatenated.

# Are there any user-facing changes?

No.

### CODE REVIEW

**Comment by alamb** [comment]
cc @wjones127  (who requested the feature) and @westonpace  who may be interested too

File: arrow-data/src/transform/mod.rs [hunk 1]
```diff
@@ -320,6 +320,11 @@ pub enum Capacities {
     /// * the capacity of the array offsets
     /// * the capacity of the child data
     List(usize, Option<Box<Capacities>>),
+    /// FixedSizeList data type
+    /// Define
+    /// * the capacity of the array
+    /// * the capacity of the child data
+    FixedSizeList(usize, Option<Box<Capacities>>),
```
Comment (tustvold) [inline arrow-data/src/transform/mod.rs:327]: I believe this is technically a breaking change, as this enumeration is not marked `non_exhaustive`. Whilst less clean perhaps we could reuse the List variant above?

**Comment by alamb** [comment]
FYI: The MSRV CI check  was fixed by @korowa  in https://github.com/apache/arrow-rs/pull/5866, so after merging up from `master` it should pass on this PR

File: arrow-data/src/transform/mod.rs [hunk 2]
```diff
@@ -506,7 +515,18 @@ impl<'a> MutableArrayData<'a> {
...
                     .collect::<Vec<_>>();
-                vec![MutableArrayData::new(children, use_nulls, array_capacity)]
+                let capacities =
+                    if let Capacities::FixedSizeList(capacity, ref child_capacities) = capacities {
+                        child_capacities
+                            .clone()
+                            .map(|c| *c)
+                            .unwrap_or(Capacities::Array(capacity))
+                    } else {
+                        Capacities::Array(array_capacity)
```
Comment (tustvold) [inline arrow-data/src/transform/mod.rs:525]: I think this should be `array_capacity * size` where `size` is from the `DataType::FixedSizeList`

Comment (judahrand) [reply arrow-data/src/transform/mod.rs]: https://github.com/apache/arrow-rs/pull/5862/commits/69c1588768668a5ec69be864ce41b956aab7a5a4

**Review by wjones127 — APPROVED** [review]
Nice work here. I see 35% improvement in the benchmark you provided, which is by no means even the worst case.


Before: `[709.23 µs 713.84 µs 719.40 µs]`
After: `[456.85 µs 461.13 µs 466.36 µs]`

**Pull request merged**